One phone number, multiple contacts

dannyroberts opened the pull request:
##### SUMMARY
Enable the use case where we have multiple contacts in our system all sharing the same phone number. (Families may share one phone for example.)

##### FEATURE FLAG
`ONE_PHONE_NUMBER_MULTIPLE_CONTACTS`


##### RISK ASSESSMENT / QA PLAN
This will for sure need good QA. I'm not quite there yet; so far I've just done a first pass, but I will think more fully about how to test this.

##### PRODUCT DESCRIPTION

dannyroberts committed:
Add ONE_PHONE_NUMBER_MULTIPLE_CONTACTS flag

dannyroberts committed:
Add XFormsSessionSynchronization to handle one-session-per-number semantics

dannyroberts committed:
Change XFormsSessionSynchronization to work off "channels"

instead of off phone numbers. We want people using the same number on different gateways
to continue to receive messages from commcare on those two different "channels" concurrently.
A "channel" is a (gateway backend, phone number) pair

dannyroberts committed:
Enable one form session at a time per channel

dannyroberts committed:
Switch Channels to use backend_id instead of backend_name

name is not unique, so that was just an oversight

dannyroberts committed:
Route incoming SMS based on the active form session for the channel

orangejenny commented on corehq/apps/smsforms/models.py:
```diff
@@ -237,3 +240,82 @@ def move_to_next_action(self):
+    @classmethod
+    def could_maybe_claim_phone_number_for_session_id(cls, phone_number, session_id):
+        """
+        Check if there's another session running for heuristic purposes
```
I'm a bit confused by the purpose of this function, what do you mean by heuristic purposes?

dannyroberts replied:
I just mean that you can use it to see if you _would be able to_ claim it right now, but that there's no guarantee that if you try to claim it, even in the next line, that this function would have predicted correctly, because it's not part of any locking that makes it atomic. The actual claim method on the other hand is atomic because of the critical section block.

So for example, in the `handle_survey_actions` command, I call this function to see if it's even worth creating a task for it. So it's a "heuristic" in the sense that it's a pretty good signal you can use to predict whether it's worth _trying_ to claim the channel for a session, but that it's only going to be accurate most of the time so it's not good for making critical decisions on. So "this said I _could_ claim the channel, so I'm going to assume I _will be able to_ claim the channel", that would be a bad assumption; whereas "this said I _could_ claim it, so I'm going to spin up a little work to try and claim it" is a good way to use it.

orangejenny replied:
Aha, thanks, that makes sense. What would you think of calling this something like `phone_number_eligible_for_session_id`?

dannyroberts replied:
The code changed since your first comment, but I took the spirit of it and renamed `could_maybe_claim_channel_for_session` to `channel_is_available_for_session`

dannyroberts replied:
1f37125

orangejenny commented on corehq/apps/smsforms/models.py:
```diff
@@ -237,3 +240,82 @@ def move_to_next_action(self):
+                running_session_info = running_session_info._replace(session_id=None)
+                cls._set_running_session_info_for_phone_number(
+                    phone_number, running_session_info,
+                    # Keep affinity for 30 days
```
Would there be a downside to keeping affinity forever?

dannyroberts replied:
Only a small downside of keeping more things in the cache. Not sure if this is the right tradeoff, but all numbers here seem arbitrary, and "forever" seems wasteful. Not sure how wasteful it would really be in practice, but that was my thinking

orangejenny replied:
Sounds reasonable. I agree all numbers seem arbitrary.

orangejenny commented on corehq/apps/smsforms/tests/test_session_synchronization.py:
```diff
@@ -0,0 +1,43 @@
+    session_a_2 = FakeSession(session_id=str(uuid4()), phone_number=phone_number_a, connection_id='Beta')
+    session_b_1 = FakeSession(session_id=str(uuid4()), phone_number=phone_number_b, connection_id='Kappa')
+
+    # Nothing set yet, so it can be claimed
```
Thanks for the comments, they're a great help.

dannyroberts replied:
You're welcome :) They're actually helpful even just for me when I'm rereading them

orangejenny commented on corehq/apps/sms/handlers/form_session.py:
```diff
@@ -11,8 +12,10 @@
+from dimagi.utils.logging import notify_error
 
 
 def form_session_handler(v, text, msg):
```
Realize you didn't introduce this, but this could be a good time to rename `v` - `verified_number`?

dannyroberts replied:
Ha, sure. Yeah I also always kind of disliked that. Um, maybe though I'd prefer to do that in a different PR just to keep this down to the minimal changes we need to make.

orangejenny commented on corehq/apps/sms/handlers/form_session.py:
```diff
@@ -23,10 +26,25 @@ def form_session_handler(v, text, msg):
+                if not session.session_is_open:
+                    # This should never happen. But if it does we should set the channel free
+                    # and act like there was no available session
+                    notify_error("The supposedly running session was a not open and was released. "
```
```suggestion
                    notify_error("The supposedly running session was not open and was released. "
```

dannyroberts replied:
🙇

orangejenny commented on corehq/apps/smsforms/models.py:
```diff
@@ -117,6 +118,8 @@ def mark_completed(self, completed):
         self.session_is_open = False
         self.completed = completed
         self.modified_time = self.end_time = utcnow()
+        if toggles.ONE_PHONE_NUMBER_MULTIPLE_CONTACTS.enabled(self.domain):
```
Does this need to be behind the flag? I don't expect people to be flipping this flag willy-nilly, but if someone turns it on briefly by mistake, it's be nice not to have open channels left around.

dannyroberts replied:
Uh, hmm. Yeah that's a good point. I approve of that reasoning :)

dannyroberts replied:
OK I thought a bit more about this, and I think right now my main concern is to have this feature fully isolated behind this flag as much as possible, so we can be as certain as possible that deploying it to production will not break any existing functionality for projects that don't turn the flag on.

I think I've convinced myself that if the toggle is not enabled, releasing or not releasing the claim on the channel has no effect. If you had a session claim that was never released because of edge cases around turning the flag on and off, and then you turn the flag on again, then that would be another story—the lingering channel claim from the finished session could prevent access for a week (i.e. until it expires). To address _that_, I've added 9506445, which has `handle_survey_actions` check for this edge case and manually release claims from closed sessions.

orangejenny commented on corehq/apps/smsforms/tasks.py:
```diff
@@ -24,6 +25,10 @@ def handle_due_survey_action(domain, contact_id, session_id):
         ):
             return
 
+        if toggles.ONE_PHONE_NUMBER_MULTIPLE_CONTACTS.enabled(domain):
```
This will cause the task to get re-queued because another contact is using it, right? A comment here would be useful.

dannyroberts replied:
Yeah it would cause it not exactly to be "requeued" but just to stay on the queue and continue to be retried regularly. Actually now that you mention it, we may need to release the enqueue lock here (or change it to hold for 1 minute instead of an hour) that is set in `handle_survey_actions`. Otherwise I think (possibly, worth looking harder at) that lock would stay in place for an hour and cause this session to get skipped in processing until the lock expired. We maybe want to try it less often than other things in the queue, but an hour is way to much of a delay I think

dannyroberts replied:
OK this should be addressed by aeaea58

orangejenny commented on corehq/apps/smsforms/models.py:
```diff
@@ -237,3 +245,101 @@ def move_to_next_action(self):
+    )
+
+
+RunningSessionInfo = namedtuple('RunningSessionInfo', ['session_id', 'contact_id'])
```
Could/should this also indicate contact type, since those can be either case ids or user ids? Or do you anticipate that it okay to assume these won't overlap, that users will always have some kind of globally unique id even when they're no longer in couch?

dannyroberts replied:
Oh, that's a good point. I'm not sure how we want to handle that, but I just knew that for the time being we didn't need it

dannyroberts replied:
I think I will ignore this avenue for now other than to look out for any place this might be an issue

orangejenny commented on corehq/apps/sms/api.py:
```diff
@@ -626,7 +627,28 @@ def get_inbound_phone_entry(msg):
+                if contact_id:
+                    p = PhoneNumber.get_phone_number_for_owner(contact_id, msg.phone_number)
+                if running_session_info.session_id and not p:
+                    # This would be very unusual, as it would meant the the supposedly running form session
```
```suggestion
                    # This would be very unusual, as it would mean the the supposedly running form session
```

orangejenny reviewed (approved):
This looks great, just left minor comments.

dannyroberts committed:
Respond to stickler

Including a broken import. Thanks stickler!

dannyroberts committed:
Update internal error text in `form_session_handler`

Co-authored-by: Jenny Schweers <orange.jenny@gmail.com>

dannyroberts committed:
Update internal error text in `get_inbound_phone_entry`

Co-authored-by: Jenny Schweers <orange.jenny@gmail.com>

mkangia commented:
loving the comments in the code @dannyroberts , can't express how much I appreciate them :)

dannyroberts committed:
Release lock on session/action when unable to claim the channel

dannyroberts committed:
Merge branch 'master' into dmr/ONE_PHONE_NUMBER_MULTIPLE_CONTACTS

dannyroberts committed:
Make explict choices around how to handle broken channels

when it comes to XFormsSessionSynchronization

dannyroberts committed:
Release any stale session/channel claims in handle_survey_action

This is a release valve for edge cases we can imagine but will almost never happen.
When a session claim is still active but the session is no longer open,
that means whatever was supposed to release it has failed to do so,
and so we release it and move on as if the channel were unclaimed.

dannyroberts committed:
Rename could_maybe_claim_channel_for_session => channel_is_available_for_session

and improve the doc string

dannyroberts committed:
Fix the way we attempt to clear a stale channel claim

Previous commit referenced undefined name "session",
but that also prompted me to rethink this.
From a robustness perspective, we want to clear the _channel_
to be reclaimed whether or not the closed session for that stale claim
still is configured to point to that channel.
Introduces XFormsSessionSynchronization.clear_stale_channel_claim
for this purpose, with tests

dannyroberts commented on corehq/apps/sms/handlers/form_session.py:
```diff
@@ -23,10 +26,25 @@ def form_session_handler(v, text, msg):
+        else:
+            multiple, session = get_single_open_session_or_close_multiple(v.domain, v.owner_id)
+            if multiple:
+                send_sms_to_verified_number(v, get_message(MSG_MULTIPLE_SESSIONS, v))
```
gonna ignore this warning (it's preexisting code that I just indented)

orangejenny reviewed (approved):
Updates look great, comments look great.

dannyroberts committed:
Rename Channel to SMSChannel

to match the namedtuple name, which is important for pickling

dannyroberts committed:
Test pickle roundtrip for SMSChannel and RunningSessionInfo

dannyroberts commented on corehq/apps/sms/handlers/form_session.py:
```diff
@@ -23,10 +26,25 @@ def form_session_handler(v, text, msg):
+        else:
+            multiple, session = get_single_open_session_or_close_multiple(v.domain, v.owner_id)
+            if multiple:
+                send_sms_to_verified_number(v, get_message(MSG_MULTIPLE_SESSIONS, v))
```
again, gonna ignore this

dannyroberts committed:
Apply ONE_PHONE_NUMBER_MULTIPLE_CONTACTS semantics when initiating an sms form

dannyroberts committed:
Merge branch 'master' into dmr/ONE_PHONE_NUMBER_MULTIPLE_CONTACTS

dannyroberts committed:
Add session_started and session_ended metrics for sms form sessions

dannyroberts committed:
Add some notify messages to send_first_message to help w/ future refactoring

dannyroberts committed:
Engage ONE_PHONE_NUMBER_MULTIPLE_CONTACTS semantics even when backend not global

dannyroberts committed:
Switch send_first_message to use background_queue

It is not strictly just for reminders, but also
  - background_queue has better lag guarantees
  - staging does not have reminder_queue which makes the feature hard to test there on that queue

dannyroberts committed:
Respond to stickler

dannyroberts committed:
Fix toggle usage

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

dannyroberts commented on corehq/apps/sms/api.py:
```diff
@@ -626,7 +627,28 @@ def get_inbound_phone_entry(msg):
         if not backend.is_global and toggles.INBOUND_SMS_LENIENCY.enabled(backend.domain):
-            p = PhoneNumber.get_two_way_number_with_domain_scope(msg.phone_number, backend.domains_with_access)
+            p = None
+            if toggles.ONE_PHONE_NUMBER_MULTIPLE_CONTACTS:
```
Eek, thanks

dannyroberts commented on corehq/apps/smsforms/models.py:
```diff
@@ -237,3 +246,158 @@ def move_to_next_action(self):
+
+
+class XFormsSessionSynchronization:
+    """
+    This class acts as a container for a set of functions
+    related to making sure each Channel(backend_id, phone_number)
+    only has one running session at a time.
+
+    Currently the backend_id `None`,
+    which represents our inability to find a suitable backend given the configuration,
+    is treated as if it were an actual backend in terms of the one-session-per-channel semantics,
+    resulting in the behavior that a phone number with a broken channel (bad PhoneNumber/backend configuration)
+    can only have one running session at a time.
+    In practice this session will fail in other ways (since there's no good channel to interact on)
+    but we let that fail downstream like it would without the synchronization piece.
+    If there seem to be backups of sessions on phone numbers with broken channels,
+    it may be worth revisiting the choice to preserve one-session-per-channel semantics even for broken channels.
+
```
This is going to come down to personal preference, but I prefer to have line breaks at natural sentence and phrase boundaries to make for nicer diffs when comments change, so no I haven't configured word wrap for my tool. If this is something the team's already hashed out and there's a word wrap guideline the team's supposed to follow, then I'm happy to of course take the team guideline over my minor personal preference

millerdev replied:
It's a [PEP-8](https://www.python.org/dev/peps/pep-0008/#maximum-line-length) thing, which I thought we'd agreed on years ago

> Some teams strongly prefer a longer line length. For code maintained exclusively or primarily by a team that can reach agreement on this issue, it is okay to increase the line length limit up to 99 characters, provided that comments and docstrings are still wrapped at 72 characters.

dannyroberts replied:
Oh, I see, the part about a shorter line length for flowing text. I wonder if we can get stickler to enforce that as well so it doesn't fall onto good souls like you :) I'll also try and set up my editor to do that for me

dannyroberts commented on corehq/apps/smsforms/models.py:
```diff
@@ -237,3 +246,158 @@ def move_to_next_action(self):
+    def _channel_affinity_cache_key(channel):
+        return f'XFormsSessionSynchronization.value.{channel.backend_id}/{channel.phone_number}'
+
+    @staticmethod
+    def _critical_section(channel):
+        return CriticalSection([
+            f'XFormsSessionSynchronization.critical_section.{channel.backend_id}/{channel.phone_number}'
+        ], timeout=5 * 60)
```
I'm not sure, would it? I thought `CriticalSection` uses redis locks, and I figured redis locks share the same key namespace as redis key/value pairs

millerdev replied:
Ah, good point. I thought `CriticalSection` applied an extra prefix or suffix on the key, but just checked and it does not. Ignore this comment.

dannyroberts committed:
Fix send_first_message refactoring notify

dannyroberts committed:
Ease definition of a "two-way" number when sharing numbers

dannyroberts committed:
Update long comment to wrap with shorter lines

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

dannyroberts committed:
Revert "Add session_started and session_ended metrics for sms form sessions"

This reverts commit 2c0b4505d2c4d3da16a333ac2bc43e676aeff14d.

dannyroberts committed:
Merge branch 'dmr/revert-commit-that-is-already-done-differently-in-another-branch' into dmr/ONE_PHONE_NUMBER_MULTIPLE_CONTACTS
